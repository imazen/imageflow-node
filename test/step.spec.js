const {
    Steps,
    FromBuffer,
    MozJPEG,
    FromURL,
    FillRect,
    BlackColor,
    Rotate90,
    Region,
    CropWhitespace,
} = require('..')

const fs = require('fs')

let str = fs.readFileSync('./test/test.jpg')

describe('basic', () => {
    it('should create a steps', async () => {
        let job = new Steps(new FromBuffer(Buffer.from('')))

        expect(typeof job).toEqual('object')
    })

    it('should be able to flip the image', async () => {
        let job = await new Steps(new FromBuffer(str))
            .flipVertical()
            .encode(new FromBuffer(null, 'key'), new MozJPEG())
            .execute()
        expect(job.key).toBeInstanceOf(Buffer)
    })

    it('should be able to constraint', async () => {
        let job = await new Steps(new FromBuffer(str))
            .constraintWithin(5, 5)
            .encode(new FromBuffer(null, 'key'), new MozJPEG())
            .execute()
        expect(job.key).toBeInstanceOf(Buffer)
    })

    it('should be able to  branch the image', async () => {
        let job = await new Steps(new FromBuffer(str))
            .branch((step) =>
                step
                    .colorFilterGrayscaleNtsc()
                    .encode(new FromBuffer(null, 'key_2'), new MozJPEG())
            )
            .encode(new FromBuffer(null, 'key'), new MozJPEG())
            .execute()
        expect(job.key).toBeInstanceOf(Buffer)
        expect(job.key_2).toBeInstanceOf(Buffer)
    })

    it('should be able to create a canvas', async () => {
        let job = await new Steps(new FromBuffer(str))
            .constraintWithin(5, 5)
            .branch((step) =>
                step
                    .drawImageExacTo(
                        (step) => step.decode(new FromBuffer(str)),
                        {
                            w: 10,
                            h: 10,
                            x: 0,
                            y: 0,
                        }
                    )
                    .encode(new FromBuffer(null, 'key_2'), new MozJPEG())
            )
            .encode(new FromBuffer(null, 'key'), new MozJPEG())
            .execute()
        expect(job.key).toBeInstanceOf(Buffer)
        expect(job.key_2).toBeInstanceOf(Buffer)
    })
})

it('should be able perform all operations', async () => {
    let job = await new Steps(new FromBuffer(str))
        .constraintWithin(100, 100)
        .distort(10, 10)
        .fillRect(new FillRect(0, 0, 8, 8, new BlackColor()))
        .flipVertical()
        .flipHorizontal()
        .rotate(new Rotate90())
        .rotate180()
        .rotate270()
        .rotate90()
        .region(new Region({ x1: 0, y1: 0, x2: 8, y2: 8 }, new BlackColor()))
        .watermark(new FromBuffer(str), {})
        .command('width=100&height=100&mode=max')
        .cropWhiteSpcae(new CropWhitespace(80, 10))
        .colorFilterAlpha(0.9)
        .colorFilterBrightness(0.8)
        .colorFilterIvert()
        .colorFilterGraycaleBt709()
        .colorFilterGrayscaleFlat()
        .colorFilterGrayscaleRY()
        .encode(new FromBuffer(null, 'key'), new MozJPEG())
        .execute()
    expect(job.key).toBeInstanceOf(Buffer)
})
