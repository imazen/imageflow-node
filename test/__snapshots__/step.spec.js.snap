// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Encode operations should create right json for toBuffer 1`] = `
Steps {
  "decodeValue": true,
  "graph": Graph {
    "_internal": Map {
      0 => Array [
        Object {
          "to": 1,
          "type": "input",
        },
      ],
      1 => Array [
        Object {
          "to": 2,
          "type": "input",
        },
      ],
      2 => Array [],
    },
  },
  "inputs": Array [
    FromFile {
      "direction": "in",
      "filename": "test_file.jpg",
      "ioId": 0,
    },
  ],
  "ioID": 2,
  "last": 2,
  "outputs": Array [
    FromBuffer {
      "buffer": null,
      "direction": "out",
      "ioId": 1,
      "key": undefined,
    },
  ],
  "vertex": Array [
    Decode {
      "commands": undefined,
      "io_id": 0,
    },
    Constrain {
      "canvasColor": null,
      "gravity": null,
      "height": 100,
      "hints": null,
      "mode": "within",
      "width": 100,
    },
    Encode {
      "io_id": 1,
      "preset": MozJPEG {
        "isProgressive": undefined,
        "matte": undefined,
        "quality": undefined,
      },
    },
  ],
}
`;

exports[`Encode operations should create right json for toFile 1`] = `
Steps {
  "decodeValue": true,
  "graph": Graph {
    "_internal": Map {
      0 => Array [
        Object {
          "to": 1,
          "type": "input",
        },
      ],
      1 => Array [
        Object {
          "to": 2,
          "type": "input",
        },
      ],
      2 => Array [],
    },
  },
  "inputs": Array [
    FromFile {
      "direction": "in",
      "filename": "test_file.jpg",
      "ioId": 0,
    },
  ],
  "ioID": 2,
  "last": 2,
  "outputs": Array [
    FromFile {
      "direction": "out",
      "filename": "test_output.jpg",
      "ioId": 1,
    },
  ],
  "vertex": Array [
    Decode {
      "commands": undefined,
      "io_id": 0,
    },
    Constrain {
      "canvasColor": null,
      "gravity": null,
      "height": 100,
      "hints": null,
      "mode": "within",
      "width": 100,
    },
    Encode {
      "io_id": 1,
      "preset": MozJPEG {
        "isProgressive": undefined,
        "matte": undefined,
        "quality": undefined,
      },
    },
  ],
}
`;

exports[`Encode operations should create right json for toJpeg 1`] = `
Steps {
  "decodeValue": true,
  "graph": Graph {
    "_internal": Map {
      0 => Array [
        Object {
          "to": 1,
          "type": "input",
        },
      ],
      1 => Array [
        Object {
          "to": 2,
          "type": "input",
        },
      ],
      2 => Array [],
    },
  },
  "inputs": Array [
    FromFile {
      "direction": "in",
      "filename": "test_file.jpg",
      "ioId": 0,
    },
  ],
  "ioID": 2,
  "last": 2,
  "outputs": Array [
    FromFile {
      "direction": "out",
      "filename": "test.jpeg",
      "ioId": 1,
    },
  ],
  "vertex": Array [
    Decode {
      "commands": undefined,
      "io_id": 0,
    },
    Constrain {
      "canvasColor": null,
      "gravity": null,
      "height": 100,
      "hints": null,
      "mode": "within",
      "width": 100,
    },
    Encode {
      "io_id": 1,
      "preset": MozJPEG {
        "isProgressive": undefined,
        "matte": undefined,
        "quality": 90,
      },
    },
  ],
}
`;

exports[`Encode operations should create right json for toPng 1`] = `
Steps {
  "decodeValue": true,
  "graph": Graph {
    "_internal": Map {
      0 => Array [
        Object {
          "to": 1,
          "type": "input",
        },
      ],
      1 => Array [
        Object {
          "to": 2,
          "type": "input",
        },
      ],
      2 => Array [],
    },
  },
  "inputs": Array [
    FromFile {
      "direction": "in",
      "filename": "test_file.jpg",
      "ioId": 0,
    },
  ],
  "ioID": 2,
  "last": 2,
  "outputs": Array [
    FromFile {
      "direction": "out",
      "filename": "test.png",
      "ioId": 1,
    },
  ],
  "vertex": Array [
    Decode {
      "commands": undefined,
      "io_id": 0,
    },
    Constrain {
      "canvasColor": null,
      "gravity": null,
      "height": 100,
      "hints": null,
      "mode": "within",
      "width": 100,
    },
    Encode {
      "io_id": 1,
      "preset": LosslessPNG {
        "maximumDeflate": false,
      },
    },
  ],
}
`;

exports[`Encode operations should create right json for toStream 1`] = `
Steps {
  "decodeValue": true,
  "graph": Graph {
    "_internal": Map {
      0 => Array [
        Object {
          "to": 1,
          "type": "input",
        },
      ],
      1 => Array [
        Object {
          "to": 2,
          "type": "input",
        },
      ],
      2 => Array [],
    },
  },
  "inputs": Array [
    FromFile {
      "direction": "in",
      "filename": "test_file.jpg",
      "ioId": 0,
    },
  ],
  "ioID": 2,
  "last": 2,
  "outputs": Array [
    FromStream {
      "direction": "out",
      "internalStream": null,
      "ioId": 1,
    },
  ],
  "vertex": Array [
    Decode {
      "commands": undefined,
      "io_id": 0,
    },
    Constrain {
      "canvasColor": null,
      "gravity": null,
      "height": 100,
      "hints": null,
      "mode": "within",
      "width": 100,
    },
    Encode {
      "io_id": 1,
      "preset": MozJPEG {
        "isProgressive": undefined,
        "matte": undefined,
        "quality": undefined,
      },
    },
  ],
}
`;

exports[`Encode operations should create right json for toWebp 1`] = `
Steps {
  "decodeValue": true,
  "graph": Graph {
    "_internal": Map {
      0 => Array [
        Object {
          "to": 1,
          "type": "input",
        },
      ],
      1 => Array [
        Object {
          "to": 2,
          "type": "input",
        },
      ],
      2 => Array [],
    },
  },
  "inputs": Array [
    FromFile {
      "direction": "in",
      "filename": "test_file.jpg",
      "ioId": 0,
    },
  ],
  "ioID": 2,
  "last": 2,
  "outputs": Array [
    FromFile {
      "direction": "out",
      "filename": "test.png",
      "ioId": 1,
    },
  ],
  "vertex": Array [
    Decode {
      "commands": undefined,
      "io_id": 0,
    },
    Constrain {
      "canvasColor": null,
      "gravity": null,
      "height": 100,
      "hints": null,
      "mode": "within",
      "width": 100,
    },
    Encode {
      "io_id": 1,
      "preset": WebPLossless {},
    },
  ],
}
`;

exports[`should be throw error if Decode options are called multiple times 1`] = `"duplicate options are not allowed"`;

exports[`should create an array for commands 1`] = `
Array [
  "ignore_color_profile_errors",
  Object {
    "jpeg_downscale_hints": Object {
      "gamma_correct_for_srgb_during_spatial_luma_scaling": true,
      "height": 100,
      "scale_luma_spatially": true,
      "width": 100,
    },
  },
  Object {
    "webp_decoder_hints": Object {
      "height": 100,
      "width": 100,
    },
  },
  "discard_color_profile",
]
`;
